app/controllers/discourse_data_explorer/query_controller.rb

```diff
@@ -135,6 +135,8 @@ def schema
     # explain - string. (Optional - pass explain=true in the request) Postgres query plan, UNIX newlines.
     # rows - array of array of strings. Results of the query. In the same order as 'columns'.
     def run
+      rate_limit_query_runs!
```

Comment:
This can raise a `RateLimiter::LimitExceeded` error which will be caught and turned into a nice JSON payload with extra headers [here](https://github.com/discourse/discourse/blob/main/app/controllers/application_controller.rb#L180-L199).